Helper/PaymentResponseHandler.php

```diff
@@ -275,17 +275,20 @@ public function handlePaymentsDetailsResponse(
             case self::CANCELLED:
                 // Cancel order in case result is refused
                 if (null !== $order) {
-                    // Set order to new so it can be cancelled
-                    $order->setState(\Magento\Sales\Model\Order::STATE_NEW);
-                    $order->save();
+                    // Check if the current state allows for changing to new for cancellation
+                    if ($order->getState() !== \Magento\Sales\Model\Order::STATE_CANCELED || \Magento\Sales\Model\Order::STATE_PENDING_PAYMENT ) {
```

Comment:
Maintaining a logic might become difficult eventually. Is it possible to use [canCancel()](https://github.com/magento/magento2/blob/e6fc0ca2d4f380951f606f15715c904b25ec9002/app/code/Magento/Sales/Model/Order.php#L633) method of `Order` class? What could be the edge cases if we use this method instead of this `if` block?